lnrpc/invoicesrpc/addinvoice.go

```diff
@@ -678,169 +627,213 @@ type SelectHopHintsCfg struct {
-		// of a private channel.
-		if channel.ScidAliasFeature {
-			var defaultScid lnwire.ShortChannelID
-			if alias == defaultScid {
-				continue
-			}
+	return privateChannels, nil
+}
+
+// shouldIncludeChannel returns true if the channel passes all the checks to
+// be a hopHint in a given invoice.
+func shouldIncludeChannel(channel *channeldb.OpenChannel,
+	alreadyIncluded map[uint64]struct{},
+	cfg *SelectHopHintsCfg) (zpay32.HopHint, lnwire.MilliSatoshi, bool) {
+
+	if _, ok := alreadyIncluded[channel.ShortChannelID.ToUint64()]; ok {
+		return zpay32.HopHint{}, 0, false
+	}
+
+	chanID := lnwire.NewChanIDFromOutPoint(
+		&channel.FundingOutpoint,
+	)
+
+	hopHintInfo := newHopHintInfo(channel, cfg.IsChannelActive(chanID))
+
+	// If this channel can't be a hop hint, then skip it.
+	edgePolicy, canBeHopHint := chanCanBeHopHint(hopHintInfo, cfg)
+	if edgePolicy == nil || !canBeHopHint {
+		return zpay32.HopHint{}, 0, false
+	}
+
+	if hopHintInfo.ScidAliasFeature {
+		alias, err := cfg.GetAlias(chanID)
+		if err != nil {
+			return zpay32.HopHint{}, 0, false
 		}
 
-		// Now that we now this channel use usable, add it as a hop
-		// hint and the indexes we'll use later.
-		addHopHint(&hopHints, channel, edgePolicy, alias)
+		_, included := alreadyIncluded[alias.ToUint64()]
+		switch {
+		case alias.IsDefault():
+			return zpay32.HopHint{}, 0, false
 
-		hopHintChans[channel.FundingOutpoint] = struct{}{}
-		totalHintBandwidth += channel.RemoteBalance
+		case included:
+			return zpay32.HopHint{}, 0, false
+
+		default:
+			hopHintInfo.ShortChannelID = alias.ToUint64()
+		}
 	}
 
-	// In this second pass we'll add channels, and we'll either stop when
-	// we have 20 hop hints, we've run through all the available channels,
-	// or if the sum of available bandwidth in the routing hints exceeds 2x
-	// the payment amount. We do 2x here to account for a margin of error
-	// if some of the selected channels no longer become operable.
-	for i := 0; i < len(openChannels); i++ {
+	// Now that we now this channel use usable, add it as a hop hint and
+	// the indexes we'll use later.
+	hopHint := newHopHint(hopHintInfo, edgePolicy)
+	return hopHint, hopHintInfo.RemoteBalance, true
+}
+
+// selectHopHints iterates a list of potential hints selecting the valid hop
+// hints until we have enough hints or run out of channels.
+//
+// NOTE: selectHopHints expects potentialHints to be already sorted in
+// descending priority.
+func selectHopHints(nHintsLeft int, targetBandwidth lnwire.MilliSatoshi,
+	potentialHints []*channeldb.OpenChannel,
+	alreadyIncluded map[uint64]struct{},
+	cfg *SelectHopHintsCfg) [][]zpay32.HopHint {
+
+	currentBandwidth := lnwire.MilliSatoshi(0)
+	hopHints := make([][]zpay32.HopHint, 0, nHintsLeft)
+	for _, channel := range potentialHints {
 		enoughHopHints := sufficientHints(
-			len(hopHints), numMaxHophints, hopHintFactor, amtMSat,
-			totalHintBandwidth,
+			nHintsLeft, currentBandwidth, targetBandwidth,
```

Comment:
can just

```
if sufficientHints(...) {
       return hopHints
}
```

Reply:
it does not fit in 80 characters, so it would look like

```
if sufficientHints(
    nHintsLeft, currentBandwidth, targetBandwidth,
) {
    return hopHints
}